File: frontend/src/api/calibration.ts
import type {
  CalibrationPointCreate,
  CalibrationStatus,
  CameraCalibrationConfig,
  FovEstimateRequest,
  FovEstimateResponse,
  DirectionCalibrationProfile,
  DirectionCalibrationStatus,
  DirectionObservationRequest,
  DirectionObservationResult,
  DirectionSimulationRequest,
  DirectionSimulationResult,
} from '../types/calibration'

function apiBaseUrl(): string {
  const configured = import.meta.env.VITE_BACKEND_API_URL as string | undefined
  if (configured) return configured.replace(/\/$/, '')
  if (window.location.port && window.location.port !== '5173') return window.location.origin
  return `${window.location.protocol}//${window.location.hostname}:8000`
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${apiBaseUrl()}${path}`, {
    headers: { 'Content-Type': 'application/json', ...init?.headers },
    ...init,
  })
  const body = await response.json()
  if (!response.ok) throw new Error(JSON.stringify(body))
  return body as T
}

export function fetchCalibrationStatus(): Promise<CalibrationStatus> {
  return request<CalibrationStatus>('/api/calibration/status')
}

export function fetchCalibrationConfig(): Promise<CameraCalibrationConfig> {
  return request<CameraCalibrationConfig>('/api/calibration/config')
}

export function updateCalibrationConfig(config: CameraCalibrationConfig): Promise<CameraCalibrationConfig> {
  return request<CameraCalibrationConfig>('/api/calibration/config', {
    method: 'PUT',
    body: JSON.stringify(config),
  })
}

export function addCalibrationPoint(point: CalibrationPointCreate): Promise<CalibrationStatus> {
  return request<CalibrationStatus>('/api/calibration/points', {
    method: 'POST',
    body: JSON.stringify(point),
  })
}

export function deleteCalibrationPoint(pointId: string): Promise<CalibrationStatus> {
  return request<CalibrationStatus>(`/api/calibration/points/${pointId}`, { method: 'DELETE' })
}

export function computeCalibration(): Promise<CalibrationStatus> {
  return request<CalibrationStatus>('/api/calibration/compute', { method: 'POST' })
}

export function resetCalibration(): Promise<CalibrationStatus> {
  return request<CalibrationStatus>('/api/calibration/reset', { method: 'POST' })
}

export function estimateFov(requestBody: FovEstimateRequest): Promise<FovEstimateResponse> {
  return request<FovEstimateResponse>('/api/calibration/fov-estimate', {
    method: 'POST',
    body: JSON.stringify(requestBody),
  })
}

export function fetchDirectionStatus(): Promise<DirectionCalibrationStatus> {
  return request<DirectionCalibrationStatus>('/api/calibration/direction/status')
}

export function simulateDirection(requestBody: DirectionSimulationRequest): Promise<DirectionSimulationResult> {
  return request<DirectionSimulationResult>('/api/calibration/direction/simulate', {
    method: 'POST',
    body: JSON.stringify(requestBody),
  })
}

export function recordDirectionObservation(requestBody: DirectionObservationRequest): Promise<DirectionObservationResult> {
  return request<DirectionObservationResult>('/api/calibration/direction/record-observation', {
    method: 'POST',
    body: JSON.stringify(requestBody),
  })
}

export function saveDirectionProfile(): Promise<DirectionCalibrationProfile> {
  return request<DirectionCalibrationProfile>('/api/calibration/direction/save-profile', { method: 'POST' })
}

export function resetDirectionProfile(): Promise<DirectionCalibrationStatus> {
  return request<DirectionCalibrationStatus>('/api/calibration/direction/reset', { method: 'POST' })
}

File: backend/app/api/calibration.py
from fastapi import APIRouter, Depends

from app.api.deps import get_runtime
from app.schemas.calibration import (
    CalibrationConfigModel,
    CalibrationPointCreate,
    CalibrationStatus,
    CameraCalibrationConfig,
    DirectionCalibrationProfile,
    DirectionCalibrationStatus,
    DirectionObservationRequest,
    DirectionObservationResult,
    DirectionSimulationRequest,
    DirectionSimulationResult,
    FovEstimateRequest,
    FovEstimateResponse,
)
from app.services.runtime_state import RuntimeState

router = APIRouter(prefix="/api/calibration", tags=["calibration"])


@router.get("/status", response_model=CalibrationStatus)
def get_calibration_status(runtime: RuntimeState = Depends(get_runtime)) -> CalibrationStatus:
    return runtime.calibration.status()


@router.get("/config", response_model=CameraCalibrationConfig)
def get_calibration_config(runtime: RuntimeState = Depends(get_runtime)) -> CameraCalibrationConfig:
    return runtime.calibration.config_model()


@router.put("/config", response_model=CameraCalibrationConfig)
def update_calibration_config(
    update: CalibrationConfigModel,
    runtime: RuntimeState = Depends(get_runtime),
) -> CameraCalibrationConfig:
    return runtime.calibration.update_config(update)


@router.post("/points", response_model=CalibrationStatus)
def add_calibration_point(
    point: CalibrationPointCreate,
    runtime: RuntimeState = Depends(get_runtime),
) -> CalibrationStatus:
    return runtime.calibration.add_point(point)


@router.delete("/points/{point_id}", response_model=CalibrationStatus)
def delete_calibration_point(
    point_id: str,
    runtime: RuntimeState = Depends(get_runtime),
) -> CalibrationStatus:
    return runtime.calibration.delete_point(point_id)


@router.post("/compute", response_model=CalibrationStatus)
def compute_calibration(runtime: RuntimeState = Depends(get_runtime)) -> CalibrationStatus:
    runtime.calibration.compute()
    return runtime.calibration.status()


@router.post("/fov-estimate", response_model=FovEstimateResponse)
def estimate_fov(
    request: FovEstimateRequest,
    runtime: RuntimeState = Depends(get_runtime),
) -> FovEstimateResponse:
    return runtime.calibration.fov_estimate(request)


@router.post("/reset", response_model=CalibrationStatus)
def reset_calibration(runtime: RuntimeState = Depends(get_runtime)) -> CalibrationStatus:
    return runtime.calibration.reset()


@router.get("/direction/status", response_model=DirectionCalibrationStatus)
def direction_status(runtime: RuntimeState = Depends(get_runtime)) -> DirectionCalibrationStatus:
    return runtime.calibration.direction_status()


@router.post("/direction/simulate", response_model=DirectionSimulationResult)
def direction_simulate(
    request: DirectionSimulationRequest,
    runtime: RuntimeState = Depends(get_runtime),
) -> DirectionSimulationResult:
    return runtime.calibration.direction_simulate(request)


@router.post("/direction/record-observation", response_model=DirectionObservationResult)
def direction_record_observation(
    request: DirectionObservationRequest,
    runtime: RuntimeState = Depends(get_runtime),
) -> DirectionObservationResult:
    return runtime.calibration.direction_record_observation(request)


@router.post("/direction/save-profile", response_model=DirectionCalibrationProfile)
def direction_save_profile(runtime: RuntimeState = Depends(get_runtime)) -> DirectionCalibrationProfile:
    return runtime.calibration.direction_save_profile()


@router.post("/direction/reset", response_model=DirectionCalibrationStatus)
def direction_reset(runtime: RuntimeState = Depends(get_runtime)) -> DirectionCalibrationStatus:
    return runtime.calibration.direction_reset()


@router.get("/direction/latest", response_model=DirectionCalibrationProfile)
def direction_latest(runtime: RuntimeState = Depends(get_runtime)) -> DirectionCalibrationProfile:
    return runtime.calibration.direction_latest()
